lib/data_layer.ex

```diff
@@ -1377,12 +1377,27 @@ defmodule AshPostgres.DataLayer do
           {upsert_field,
            Ecto.Query.dynamic(
              [],
-             fragment("EXCLUDED.?", literal(^to_string(upsert_field)))
+             fragment(
+               "EXCLUDED.?",
+               literal(^to_string(get_source_for_upsert_field(upsert_field, resource)))
+             )
            )}
       end
     end)
   end
 
+  defp get_source_for_upsert_field(field, resource) do
+    case resource
+         |> Ash.Resource.Info.attributes()
+         |> Enum.find(&(&1.name == field)) do
+      %{source: source} when not is_nil(source) ->
+        source
+
+      _ ->
+        field
+    end
+  end
```

Comment:
Is there a better way to get the argument than looking through all the attributes?

Reply:
`Ash.Resource.Info.attribute(resource, field)` will get it 👍